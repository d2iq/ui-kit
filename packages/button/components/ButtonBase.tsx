import * as React from "react";
import { cx } from "@emotion/css";
import FocusStyleManager from "../../shared/components/FocusStyleManager";
import { IconShapes } from "../../icon/components/Icon";
import IconPropAdapter from "../../icon/components/IconPropAdapter";
import {
  buttonReset,
  textWeight,
  flex,
  flexItem,
  display,
  padding
} from "../../shared/styles/styleUtils";
import {
  button,
  buttonBase,
  buttonInverse,
  fullWidthButton,
  focusStyleByAppearance,
  getInverseMutedButtonStyles,
  getMutedButtonStyles,
  processingTextStyle
} from "../style";
import { LinkProps } from "../../link/types";
import UnstyledLink from "../../link/components/UnstyledLink";

export enum ButtonAppearances {
  Primary = "primary",
  Secondary = "secondary",
  Standard = "standard",
  Danger = "danger",
  Success = "success"
}

export interface ButtonProps extends LinkProps {
  /**
   * if the button triggers new content to appear (e.g.: modals and dropdowns)
   */
  ariaHaspopup?: boolean;
  /**
   * should be used if the button does not contain text children. For example: a button that is just an icon
   */
  ariaLabel?: string;
  children?: React.ReactNode | string;
  /**
   * whether or not the button is enabled
   */
  disabled?: boolean;
  /**
   * the icon that appears before the button text
   */
  // TODO: only accept IconShapes when we make a big breaking change
  iconStart?: IconShapes | React.ReactElement<HTMLElement>;
  /**
   * the icon that appears after the button text
   */
  // TODO: only accept IconShapes when we make a big breaking change
  iconEnd?: IconShapes | React.ReactElement<HTMLElement>;
  /**
   * whether the action the button was meant to do has completed
   */
  isProcessing?: boolean;
  /**
   * if the button should fill the width of it's container
   */
  isFullWidth?: boolean;
  /**
   * if the button is on a dark background
   */
  isInverse?: boolean;
  /**
   * the funtion that is called when the button loses focus
   */
  onBlur?: (e?: React.SyntheticEvent<HTMLElement>) => void;
  /**
   * the funtion that is called when the button gets focus
   */
  onFocus?: (e?: React.SyntheticEvent<HTMLElement>) => void;
  /**
   * the funtion that is called when the button is "clicked" via cursor, touch, or keyboard
   */
  onClick?: (e?: React.SyntheticEvent<HTMLElement>) => void;
  /**
   * the type of button - same as HTML "type" attribute on the <button> tag
   */
  type?: "button" | "reset" | "submit";
}

export interface ButtonBaseProps extends ButtonProps {
  appearance: ButtonAppearances;
  className?: string;
}

class ButtonBase extends React.PureComponent<ButtonBaseProps, {}> {
  constructor(props) {
    super(props);
    this.onClick = this.onClick.bind(this);
    this.getButtonContent = this.getButtonContent.bind(this);
  }

  public render() {
    const {
      appearance,
      children,
      className,
      disabled,
      iconStart,
      iconEnd,
      isInverse,
      isProcessing,
      isFullWidth,
      onClick,
      type = "button",
      url,
      ...other
    } = this.props;

    const buttonClassName = cx(
      buttonReset,
      button(appearance),
      buttonBase,
      textWeight("medium"),
      className,
      {
        [fullWidthButton]: isFullWidth,
        [buttonInverse(appearance)]: isInverse,
        [getMutedButtonStyles(appearance)]: disabled || isProcessing,
        [getInverseMutedButtonStyles(appearance)]:
          (disabled || isProcessing) && isInverse
      }
    );

    const getButtonNode = () => {
      if (url) {
        return !disabled && !isProcessing ? (
          <UnstyledLink
            href={url}
            className={buttonClassName}
            onClick={this.onClick}
            tabIndex={0}
            {...other}
          >
            {this.getButtonContent()}
          </UnstyledLink>
        ) : (
          <UnstyledLink
            className={buttonClassName}
            aria-disabled="true"
            tabIndex={-1}
            {...other}
          >
            {this.getButtonContent()}
          </UnstyledLink>
        );
      }

      return (
        <button
          className={buttonClassName}
          disabled={disabled || isProcessing}
          onClick={this.onClick}
          tabIndex={0}
          type={type}
          {...other}
        >
          {this.getButtonContent()}
        </button>
      );
    };

    return (
      <FocusStyleManager
        focusEnabledClass={focusStyleByAppearance(appearance, isInverse)}
      >
        {getButtonNode()}
      </FocusStyleManager>
    );
  }

  private onClick(e: React.SyntheticEvent<HTMLElement>) {
    if (!this.props.disabled && this.props.onClick) {
      this.props.onClick(e);
    }
  }

  private getIconStart(icon) {
    return (
      <span className={cx(flexItem("shrink"), display("inherit"))}>
        <IconPropAdapter icon={icon} size="xs" color="inherit" />
      </span>
    );
  }

  private getIconEnd(icon) {
    return (
      Boolean(icon) && (
        <span
          className={cx(
            flexItem("shrink"),
            display("inherit"),
            padding("left", "xxs")
          )}
        >
          <IconPropAdapter icon={icon} size="xs" color="inherit" />
        </span>
      )
    );
  }

  private getButtonContent() {
    const { iconStart, iconEnd, isProcessing, children } = this.props;

    return Boolean(iconStart) || Boolean(iconEnd) ? (
      <span className={flex({ align: "center", justify: "center" })}>
        {iconStart && this.getIconStart(iconStart)}
        {Boolean(children) && (
          <span
            className={cx(flexItem("shrink"), padding("left", "xs"), {
              [processingTextStyle]: isProcessing
            })}
          >
            {children}
          </span>
        )}
        {iconEnd && this.getIconEnd(iconEnd)}
      </span>
    ) : (
      <span className={isProcessing ? processingTextStyle : ""}>
        {children}
      </span>
    );
  }
}

export default ButtonBase;
