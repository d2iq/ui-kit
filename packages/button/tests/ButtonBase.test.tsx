import React from "react";

import ButtonBase, { ButtonAppearances } from "../components/ButtonBase";
import { SystemIcons } from "../../icons/dist/system-icons-enum";
import serializer from "@emotion/jest";
import { shallow, mount } from "enzyme";
import toJson from "enzyme-to-json";

expect.addSnapshotSerializer(serializer);

describe("ButtonBase", () => {
  it("renders all appearances with props", () => {
    Object.keys(ButtonAppearances).forEach(appearance => {
      const component = shallow(
        <ButtonBase
          appearance={ButtonAppearances[appearance]}
          isFullWidth={true}
          type="submit"
          ariaHaspopup={true}
          disabled={true}
          isProcessing={true}
          iconStart={SystemIcons.Check}
          iconEnd={SystemIcons.Check}
        >
          Button
        </ButtonBase>
      );
      expect(toJson(component)).toMatchSnapshot();
    });
  });
  it("calls onClick prop when clicked", () => {
    const someFn = jest.fn();
    const component = mount(
      <ButtonBase appearance={ButtonAppearances.Standard} onClick={someFn}>
        Button
      </ButtonBase>
    );
    expect(someFn).not.toHaveBeenCalled();
    component.simulate("click");
    expect(someFn).toHaveBeenCalled();
  });
  it("does not call onClick prop when disabled", () => {
    const anotherFn = jest.fn();
    const component = shallow(
      <ButtonBase
        appearance={ButtonAppearances.Standard}
        disabled={true}
        onClick={anotherFn}
      >
        Button
      </ButtonBase>
    );
    expect(anotherFn).not.toHaveBeenCalled();
    component.simulate("click");
    expect(anotherFn).not.toHaveBeenCalled();
  });
});
