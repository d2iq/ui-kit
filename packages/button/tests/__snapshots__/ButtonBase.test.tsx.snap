// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ButtonBase renders all appearances with props 1`] = `
.emotion-0 {
  background: none;
  border: 0;
  color: inherit;
  font: inherit;
  line-height: normal;
  overflow: visible;
  padding: 0;
  text-align: inherit;
  color: #FFFFFF;
  fill: #FFFFFF;
  background-color: #7D58FF;
  border-radius: 4px;
  padding: 10px 18px;
  cursor: pointer;
  display: inline-block;
  outline: none;
  -webkit-text-decoration: none;
  text-decoration: none;
  font-weight: 500;
  text-align: center;
  width: 100%;
  color: var(--themeTextColorDisabled, #AEB0B4);
  fill: var(--themeTextColorDisabled, #AEB0B4);
  cursor: default;
  pointer-events: none;
  background-color: var(--themeBgDisabled, #E8EAED);
}

.emotion-0::-moz-focus-inner {
  border: 0;
  padding: 0;
}

.emotion-0[href],
.emotion-0[href]:visited {
  color: #FFFFFF;
  fill: #FFFFFF;
}

.emotion-0:hover {
  background-color: #704fe5;
}

.emotion-0:active {
  background-color: #6446cc;
}

.emotion-0:hover,
.emotion-0:focus,
.emotion-0:active {
  color: var(--themeTextColorDisabled, #AEB0B4);
  fill: var(--themeTextColorDisabled, #AEB0B4);
}

.emotion-0:hover,
.emotion-0:focus,
.emotion-0:active {
  background-color: var(--themeBgDisabled, #E8EAED);
}

.emotion-1 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  height: auto;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  min-height: 0;
}

.emotion-1>div {
  width: auto;
}

.emotion-2 {
  box-sizing: border-box;
  -webkit-flex-basis: auto;
  -ms-flex-preferred-size: auto;
  flex-basis: auto;
  -webkit-box-flex: 0;
  -webkit-flex-grow: 0;
  -ms-flex-positive: 0;
  flex-grow: 0;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  width: initial;
  display: inherit;
}

.emotion-3 {
  box-sizing: border-box;
  -webkit-flex-basis: auto;
  -ms-flex-preferred-size: auto;
  flex-basis: auto;
  -webkit-box-flex: 0;
  -webkit-flex-grow: 0;
  -ms-flex-positive: 0;
  flex-grow: 0;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  width: initial;
  padding-left: 8px;
}

.emotion-3:after {
  content: "...";
}

.emotion-4 {
  box-sizing: border-box;
  -webkit-flex-basis: auto;
  -ms-flex-preferred-size: auto;
  flex-basis: auto;
  -webkit-box-flex: 0;
  -webkit-flex-grow: 0;
  -ms-flex-positive: 0;
  flex-grow: 0;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  width: initial;
  display: inherit;
  padding-left: 4px;
}

<FocusStyleManager
  focusEnabledClass="css-xcotv6"
>
  <button
    ariaHaspopup={true}
    className="emotion-0"
    disabled={true}
    onClick={[Function]}
    tabIndex={0}
    type="submit"
  >
    <span
      className="emotion-1"
    >
      <span
        className="emotion-2"
      >
        <IconPropAdapter
          color="inherit"
          icon="system-check"
          size="xs"
        />
      </span>
      <span
        className="emotion-3"
      >
        Button
      </span>
      <span
        className="emotion-4"
      >
        <IconPropAdapter
          color="inherit"
          icon="system-check"
          size="xs"
        />
      </span>
    </span>
  </button>
</FocusStyleManager>
`;

exports[`ButtonBase renders all appearances with props 2`] = `
.emotion-0 {
  background: none;
  border: 0;
  color: inherit;
  font: inherit;
  line-height: normal;
  overflow: visible;
  padding: 0;
  text-align: inherit;
  color: var(--themeTextColorInteractive, #7D58FF);
  fill: var(--themeTextColorInteractive, #7D58FF);
  cursor: pointer;
  display: inline-block;
  outline: none;
  -webkit-text-decoration: none;
  text-decoration: none;
  font-weight: 500;
  text-align: center;
  width: 100%;
  color: var(--themeTextColorDisabled, #AEB0B4);
  fill: var(--themeTextColorDisabled, #AEB0B4);
  cursor: default;
  pointer-events: none;
}

.emotion-0::-moz-focus-inner {
  border: 0;
  padding: 0;
}

.emotion-0:hover {
  color: #704fe5;
  fill: #704fe5;
}

.emotion-0:active {
  color: #6446cc;
  fill: #6446cc;
}

.emotion-0[href],
.emotion-0[href]:visited {
  color: var(--themeTextColorInteractive, #7D58FF);
  fill: var(--themeTextColorInteractive, #7D58FF);
}

.emotion-0:hover,
.emotion-0:focus,
.emotion-0:active {
  color: var(--themeTextColorDisabled, #AEB0B4);
  fill: var(--themeTextColorDisabled, #AEB0B4);
}

.emotion-1 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  height: auto;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  min-height: 0;
}

.emotion-1>div {
  width: auto;
}

.emotion-2 {
  box-sizing: border-box;
  -webkit-flex-basis: auto;
  -ms-flex-preferred-size: auto;
  flex-basis: auto;
  -webkit-box-flex: 0;
  -webkit-flex-grow: 0;
  -ms-flex-positive: 0;
  flex-grow: 0;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  width: initial;
  display: inherit;
}

.emotion-3 {
  box-sizing: border-box;
  -webkit-flex-basis: auto;
  -ms-flex-preferred-size: auto;
  flex-basis: auto;
  -webkit-box-flex: 0;
  -webkit-flex-grow: 0;
  -ms-flex-positive: 0;
  flex-grow: 0;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  width: initial;
  padding-left: 8px;
}

.emotion-3:after {
  content: "...";
}

.emotion-4 {
  box-sizing: border-box;
  -webkit-flex-basis: auto;
  -ms-flex-preferred-size: auto;
  flex-basis: auto;
  -webkit-box-flex: 0;
  -webkit-flex-grow: 0;
  -ms-flex-positive: 0;
  flex-grow: 0;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  width: initial;
  display: inherit;
  padding-left: 4px;
}

<FocusStyleManager
  focusEnabledClass="css-1h85cd2"
>
  <button
    ariaHaspopup={true}
    className="emotion-0"
    disabled={true}
    onClick={[Function]}
    tabIndex={0}
    type="submit"
  >
    <span
      className="emotion-1"
    >
      <span
        className="emotion-2"
      >
        <IconPropAdapter
          color="inherit"
          icon="system-check"
          size="xs"
        />
      </span>
      <span
        className="emotion-3"
      >
        Button
      </span>
      <span
        className="emotion-4"
      >
        <IconPropAdapter
          color="inherit"
          icon="system-check"
          size="xs"
        />
      </span>
    </span>
  </button>
</FocusStyleManager>
`;

exports[`ButtonBase renders all appearances with props 3`] = `
.emotion-0 {
  background: none;
  border: 0;
  color: inherit;
  font: inherit;
  line-height: normal;
  overflow: visible;
  padding: 0;
  text-align: inherit;
  color: #1B2029;
  fill: #1B2029;
  background-color: #DADDE2;
  border-radius: 4px;
  padding: 10px 18px;
  cursor: pointer;
  display: inline-block;
  outline: none;
  -webkit-text-decoration: none;
  text-decoration: none;
  font-weight: 500;
  text-align: center;
  width: 100%;
  color: var(--themeTextColorDisabled, #AEB0B4);
  fill: var(--themeTextColorDisabled, #AEB0B4);
  cursor: default;
  pointer-events: none;
  background-color: var(--themeBgDisabled, #E8EAED);
}

.emotion-0::-moz-focus-inner {
  border: 0;
  padding: 0;
}

.emotion-0[href],
.emotion-0[href]:visited {
  color: #1B2029;
  fill: #1B2029;
}

.emotion-0:hover {
  background-color: #c4c6cb;
}

.emotion-0:active {
  background-color: #aeb0b4;
}

.emotion-0:hover,
.emotion-0:focus,
.emotion-0:active {
  color: var(--themeTextColorDisabled, #AEB0B4);
  fill: var(--themeTextColorDisabled, #AEB0B4);
}

.emotion-0:hover,
.emotion-0:focus,
.emotion-0:active {
  background-color: var(--themeBgDisabled, #E8EAED);
}

.emotion-1 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  height: auto;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  min-height: 0;
}

.emotion-1>div {
  width: auto;
}

.emotion-2 {
  box-sizing: border-box;
  -webkit-flex-basis: auto;
  -ms-flex-preferred-size: auto;
  flex-basis: auto;
  -webkit-box-flex: 0;
  -webkit-flex-grow: 0;
  -ms-flex-positive: 0;
  flex-grow: 0;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  width: initial;
  display: inherit;
}

.emotion-3 {
  box-sizing: border-box;
  -webkit-flex-basis: auto;
  -ms-flex-preferred-size: auto;
  flex-basis: auto;
  -webkit-box-flex: 0;
  -webkit-flex-grow: 0;
  -ms-flex-positive: 0;
  flex-grow: 0;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  width: initial;
  padding-left: 8px;
}

.emotion-3:after {
  content: "...";
}

.emotion-4 {
  box-sizing: border-box;
  -webkit-flex-basis: auto;
  -ms-flex-preferred-size: auto;
  flex-basis: auto;
  -webkit-box-flex: 0;
  -webkit-flex-grow: 0;
  -ms-flex-positive: 0;
  flex-grow: 0;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  width: initial;
  display: inherit;
  padding-left: 4px;
}

<FocusStyleManager
  focusEnabledClass="css-bwvukd"
>
  <button
    ariaHaspopup={true}
    className="emotion-0"
    disabled={true}
    onClick={[Function]}
    tabIndex={0}
    type="submit"
  >
    <span
      className="emotion-1"
    >
      <span
        className="emotion-2"
      >
        <IconPropAdapter
          color="inherit"
          icon="system-check"
          size="xs"
        />
      </span>
      <span
        className="emotion-3"
      >
        Button
      </span>
      <span
        className="emotion-4"
      >
        <IconPropAdapter
          color="inherit"
          icon="system-check"
          size="xs"
        />
      </span>
    </span>
  </button>
</FocusStyleManager>
`;

exports[`ButtonBase renders all appearances with props 4`] = `
.emotion-0 {
  background: none;
  border: 0;
  color: inherit;
  font: inherit;
  line-height: normal;
  overflow: visible;
  padding: 0;
  text-align: inherit;
  color: #FFFFFF;
  fill: #FFFFFF;
  background-color: #EB293A;
  border-radius: 4px;
  padding: 10px 18px;
  cursor: pointer;
  display: inline-block;
  outline: none;
  -webkit-text-decoration: none;
  text-decoration: none;
  font-weight: 500;
  text-align: center;
  width: 100%;
  color: var(--themeTextColorDisabled, #AEB0B4);
  fill: var(--themeTextColorDisabled, #AEB0B4);
  cursor: default;
  pointer-events: none;
  background-color: var(--themeBgDisabled, #E8EAED);
}

.emotion-0::-moz-focus-inner {
  border: 0;
  padding: 0;
}

.emotion-0[href],
.emotion-0[href]:visited {
  color: #FFFFFF;
  fill: #FFFFFF;
}

.emotion-0:hover {
  background-color: #d32434;
}

.emotion-0:active {
  background-color: #bc202e;
}

.emotion-0:hover,
.emotion-0:focus,
.emotion-0:active {
  color: var(--themeTextColorDisabled, #AEB0B4);
  fill: var(--themeTextColorDisabled, #AEB0B4);
}

.emotion-0:hover,
.emotion-0:focus,
.emotion-0:active {
  background-color: var(--themeBgDisabled, #E8EAED);
}

.emotion-1 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  height: auto;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  min-height: 0;
}

.emotion-1>div {
  width: auto;
}

.emotion-2 {
  box-sizing: border-box;
  -webkit-flex-basis: auto;
  -ms-flex-preferred-size: auto;
  flex-basis: auto;
  -webkit-box-flex: 0;
  -webkit-flex-grow: 0;
  -ms-flex-positive: 0;
  flex-grow: 0;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  width: initial;
  display: inherit;
}

.emotion-3 {
  box-sizing: border-box;
  -webkit-flex-basis: auto;
  -ms-flex-preferred-size: auto;
  flex-basis: auto;
  -webkit-box-flex: 0;
  -webkit-flex-grow: 0;
  -ms-flex-positive: 0;
  flex-grow: 0;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  width: initial;
  padding-left: 8px;
}

.emotion-3:after {
  content: "...";
}

.emotion-4 {
  box-sizing: border-box;
  -webkit-flex-basis: auto;
  -ms-flex-preferred-size: auto;
  flex-basis: auto;
  -webkit-box-flex: 0;
  -webkit-flex-grow: 0;
  -ms-flex-positive: 0;
  flex-grow: 0;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  width: initial;
  display: inherit;
  padding-left: 4px;
}

<FocusStyleManager
  focusEnabledClass="css-txv5vz"
>
  <button
    ariaHaspopup={true}
    className="emotion-0"
    disabled={true}
    onClick={[Function]}
    tabIndex={0}
    type="submit"
  >
    <span
      className="emotion-1"
    >
      <span
        className="emotion-2"
      >
        <IconPropAdapter
          color="inherit"
          icon="system-check"
          size="xs"
        />
      </span>
      <span
        className="emotion-3"
      >
        Button
      </span>
      <span
        className="emotion-4"
      >
        <IconPropAdapter
          color="inherit"
          icon="system-check"
          size="xs"
        />
      </span>
    </span>
  </button>
</FocusStyleManager>
`;

exports[`ButtonBase renders all appearances with props 5`] = `
.emotion-0 {
  background: none;
  border: 0;
  color: inherit;
  font: inherit;
  line-height: normal;
  overflow: visible;
  padding: 0;
  text-align: inherit;
  color: #FFFFFF;
  fill: #FFFFFF;
  background-color: #14C684;
  border-radius: 4px;
  padding: 10px 18px;
  cursor: pointer;
  display: inline-block;
  outline: none;
  -webkit-text-decoration: none;
  text-decoration: none;
  font-weight: 500;
  text-align: center;
  width: 100%;
  color: var(--themeTextColorDisabled, #AEB0B4);
  fill: var(--themeTextColorDisabled, #AEB0B4);
  cursor: default;
  pointer-events: none;
  background-color: var(--themeBgDisabled, #E8EAED);
}

.emotion-0::-moz-focus-inner {
  border: 0;
  padding: 0;
}

.emotion-0[href],
.emotion-0[href]:visited {
  color: #FFFFFF;
  fill: #FFFFFF;
}

.emotion-0:hover {
  background-color: #12b276;
}

.emotion-0:active {
  background-color: #109e69;
}

.emotion-0:hover,
.emotion-0:focus,
.emotion-0:active {
  color: var(--themeTextColorDisabled, #AEB0B4);
  fill: var(--themeTextColorDisabled, #AEB0B4);
}

.emotion-0:hover,
.emotion-0:focus,
.emotion-0:active {
  background-color: var(--themeBgDisabled, #E8EAED);
}

.emotion-1 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  height: auto;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  min-height: 0;
}

.emotion-1>div {
  width: auto;
}

.emotion-2 {
  box-sizing: border-box;
  -webkit-flex-basis: auto;
  -ms-flex-preferred-size: auto;
  flex-basis: auto;
  -webkit-box-flex: 0;
  -webkit-flex-grow: 0;
  -ms-flex-positive: 0;
  flex-grow: 0;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  width: initial;
  display: inherit;
}

.emotion-3 {
  box-sizing: border-box;
  -webkit-flex-basis: auto;
  -ms-flex-preferred-size: auto;
  flex-basis: auto;
  -webkit-box-flex: 0;
  -webkit-flex-grow: 0;
  -ms-flex-positive: 0;
  flex-grow: 0;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  width: initial;
  padding-left: 8px;
}

.emotion-3:after {
  content: "...";
}

.emotion-4 {
  box-sizing: border-box;
  -webkit-flex-basis: auto;
  -ms-flex-preferred-size: auto;
  flex-basis: auto;
  -webkit-box-flex: 0;
  -webkit-flex-grow: 0;
  -ms-flex-positive: 0;
  flex-grow: 0;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  width: initial;
  display: inherit;
  padding-left: 4px;
}

<FocusStyleManager
  focusEnabledClass="css-u3f3k6"
>
  <button
    ariaHaspopup={true}
    className="emotion-0"
    disabled={true}
    onClick={[Function]}
    tabIndex={0}
    type="submit"
  >
    <span
      className="emotion-1"
    >
      <span
        className="emotion-2"
      >
        <IconPropAdapter
          color="inherit"
          icon="system-check"
          size="xs"
        />
      </span>
      <span
        className="emotion-3"
      >
        Button
      </span>
      <span
        className="emotion-4"
      >
        <IconPropAdapter
          color="inherit"
          icon="system-check"
          size="xs"
        />
      </span>
    </span>
  </button>
</FocusStyleManager>
`;
